### DISCUSSION TITLE
Use conventional AWS environment variables for S3 credentials

### DISCUSSION THREAD

**Discussion by virtual-machinist** [afs/rdf-delta#267] [Ideas]
Hi!

Is it possible to make use of the standard AWS environment variables (`AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`) for S3 config instead of the credentials file option?

When running in Docker I can work around this with a custom entrypoint script that creates the credentials file before starting the server, but wouldn't it be better to simply delegate to something like
```
new AWSCredentialsProviderChain(
              new EnvironmentVariableCredentialsProvider(),
              new AWSStaticCredentialsProvider(new AnonymousAWSCredentials())
)
```
instead of static anonymous credentials.

If we add a `new ProfileCredentialsProvider()` before the anonymous credentials provider then there's even no need to provide a special S3 credentials file command line parameter as it may be fetched by the AWS library using conventional options AFAICT.

Or am I missing something and there's another way to use the environment variables already?

**Comment by afs** [comment]
Hi @virtual-machinist - have you tried no providing the keys file? From reading the code, (I don't have a setup to hand just at the moment) it defaults to standard AWS S3 setup, but maybe some earlier checking happens to require the keys file.

See `S3.buildS3` . The AWS setup starts with `AmazonS3ClientBuilder.standard()` which has `new S3CredentialsProviderChain()`

**Reply by virtual-machinist** [reply]
Running `dcmd` with the `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` environment variables, i.e. without the profile file was the first thing I've tried, but got a 403.

You're correct in the sense that `AmazonS3ClientBuilder.standard()` does have the `new S3CredentialsProviderChain()`.
However the next statements are as follows:
```java
        if ( endpoint == null )
            builder.withRegion(region);
        else  {
            // Needed for S3mock
            builder.withPathStyleAccessEnabled(true);
            builder.withEndpointConfiguration(new EndpointConfiguration(endpoint, region));
            builder.withCredentials(new AWSStaticCredentialsProvider(new AnonymousAWSCredentials()));
        }
        if ( credentialsFile != null )
            builder.withCredentials(new ProfileCredentialsProvider(credentialsFile, credentialsProfile));
        return builder.build();
```
The `com.amazonaws.client.builder.AwsClientBuilder#withCredentials` is implemented as follows:
```java
    public final Subclass withCredentials(AWSCredentialsProvider credentialsProvider) {
        setCredentials(credentialsProvider);
        return getSubclass();
    }
```
meaning that the default credentials provider is overridden either with anonymous credentials or the profile credentials.

I do use an S3-compatible object storage, i.e. not AWS, so my `endpoint` is never `null`. I'm guessing that's why the default credentials provider gets overridden.

**Reply by afs** [reply]
> I do use an S3-compatible object storage

Which one?

**Reply by virtual-machinist** [reply]
Locally MinIO, but the project I'm trying to use RDF Delta in uses some proprietary solution, I've no way to tell what exactly.
The thing I haven't tried yet is setting up environment variables for everything related to S3 besides the required bucket name and region. That may help since then the `endpoint` can be `null`.

**Reply by virtual-machinist** [reply]
Nope, neither `AWS_ENDPOINT_URL` nor `AWS_ENDPOINT_URL_S3` work, the client tries accessing AWS. Apparently setting service-specific endpoints [is only supported](https://docs.aws.amazon.com/sdkref/latest/guide/feature-ss-endpoints.html) for AWS CLI that I'm using elsewhere, not the Java client used here.

**Reply by virtual-machinist** [reply]
@afs what do you think, can this be implemented in any way? Can do a PR with my suggested change if that's an acceptable solution.

**Reply by afs** [reply]
@virtual-machinist yes - a PR to change that code would be great to see. 

Would defaulting to "normal" (whatever AWS standard setup is) and let a file override that work? (I'm not a MinIO user.)

**Reply by virtual-machinist** [reply]
@afs I added the environment variable credentials provider with the fallback to anonymous credentials to minimize possible breakage. See #272 .
